File: app/src/main/java/com/example/cashflow/Posizione.java
package com.example.cashflow;

import android.content.pm.PackageManager;
import android.location.Address;
import android.location.Geocoder;
import android.location.Location;

import android.content.Context;
import android.Manifest;
import android.net.DnsResolver;

import androidx.core.app.ActivityCompat;

import com.example.cashflow.dataClass.City;
import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationServices;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Tasks;

import java.io.IOException;
import java.util.List;
import java.util.Locale;
import java.util.concurrent.ExecutionException;

public class Posizione {
    private FusedLocationProviderClient fusedLocationProviderClient;
    private Context context;

    public Posizione(Context context) {
        this.context = context;
        fusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(context);
    }

    public void requestDeviceLocation(final DeviceLocationCallback callback) {
        if (ActivityCompat.checkSelfPermission(context, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
            fusedLocationProviderClient.getLastLocation()
                    .addOnSuccessListener(new OnSuccessListener<Location>() {
                        @Override
                        public void onSuccess(Location location) {
                            if (location != null) {
                                float latitude = (float) location.getLatitude();
                                System.out.println("location.getLatitude() " + location.getLatitude());
                                System.out.println("latitude "+ latitude);
                                float longitude = (float) location.getLongitude();
                                Geocoder geocoder = new Geocoder(context, Locale.getDefault());
                                try {
                                    List<Address> addresses = geocoder.getFromLocation(latitude, longitude, 1);
                                    if (addresses != null && addresses.size() > 0) {
                                        String cityName = addresses.get(0).getLocality();
                                        System.out.println("cityName " + cityName);
                                        callback.onLocationFetched(new City(cityName, latitude, longitude));
                                    }
                                } catch (IOException e) {
                                    e.printStackTrace();
                                    callback.onLocationFetchFailed(e);
                                }
                            } else {
                                callback.onLocationFetchFailed(new Exception("Location is null"));
                            }
                        }
                    });
        } else {
            callback.onLocationFetchFailed(new Exception("Permission not granted"));
        }
    }

    public interface DeviceLocationCallback {
        void onLocationFetched(City city);
        void onLocationFetchFailed(Exception e);
    }
}

File: app/src/main/java/com/example/cashflow/Posizione.kt
package com.example.cashflow

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.location.Geocoder
import android.location.Location
import androidx.core.app.ActivityCompat
import com.google.android.gms.location.FusedLocationProviderClient
import com.google.android.gms.location.LocationServices
import java.io.IOException
import java.util.Locale
import com.example.cashflow.dataClass.*

class Posizione(private val context: Context) {
    private val fusedLocationProviderClient: FusedLocationProviderClient

    init {
        fusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(context)
    }

    fun requestDeviceLocation(callback: DeviceLocationCallback) {
        if (ActivityCompat.checkSelfPermission(
                context,
                Manifest.permission.ACCESS_FINE_LOCATION
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            callback.onLocationFetchFailed(Exception("Permission not granted"))
            return
        }
        fusedLocationProviderClient.lastLocation
            .addOnSuccessListener { location: Location? ->
                if (location == null) {
                    callback.onLocationFetchFailed(Exception("Location is null"))
                    return@addOnSuccessListener
                }
                val latitude = location.latitude
                val longitude = location.longitude
                val geocoder = Geocoder(context, Locale.getDefault())
                try {
                    val addresses =
                        geocoder.getFromLocation(latitude.toDouble(), longitude.toDouble(), 1)
                    if (addresses != null && !addresses.isEmpty()) {
                        val cityName = addresses[0].locality
                        callback.onLocationFetched(City(cityName, latitude, longitude))
                    }
                } catch (e: IOException) {
                    callback.onLocationFetchFailed(e)
                }
            }
    }

    interface DeviceLocationCallback {
        fun onLocationFetched(city: City?)
        fun onLocationFetchFailed(e: Exception?)
    }
}